File: src/test/java/com/uwetrottmann/trakt5/services/CommentsTest.java
/*
 * Copyright 2024 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.uwetrottmann.trakt5.services;

import com.uwetrottmann.trakt5.BaseTestCase;
import com.uwetrottmann.trakt5.TestData;
import com.uwetrottmann.trakt5.entities.Comment;
import com.uwetrottmann.trakt5.entities.Episode;
import com.uwetrottmann.trakt5.entities.EpisodeIds;
import org.junit.Test;
import retrofit2.Response;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class CommentsTest extends BaseTestCase {

    @Test
    public void test_postAndUpdate() throws InterruptedException, IOException {
        // first post a new comment
        Comment comment = new Comment(buildTestEpisode(), "This is a toasty comment!", true, false);
        Comment commentResponse = executeCall(getTrakt().comments().post(comment));

        // give the server some time to handle the data
        Thread.sleep(3000);

        // update the new comment
        Comment updatedComment = new Comment("This is toasty! I was just updated.", false, false);
        commentResponse = executeCall(getTrakt().comments().update(commentResponse.id, updatedComment));
        assertCommentResponse(updatedComment, commentResponse);
        // give the server some time to handle the data
        Thread.sleep(3000);

        // delete the comment again
        Response<Void> response = getTrakt().comments().delete(commentResponse.id).execute();
        assertSuccessfulResponse(response);
        assertThat(response.code()).isEqualTo(HttpURLConnection.HTTP_NO_CONTENT);
    }

    @Test
    public void test_delete() throws InterruptedException, IOException {
        // first post a new comment
        Comment comment = new Comment(buildTestEpisode(), "This is toasty! I should be deleted soon.", true, false);
        Comment commentResponse = executeCall(getTrakt().comments().post(comment));

        // give the server some time to handle the data
        Thread.sleep(3000);

        // delete the comment again
        Response<Void> response = getTrakt().comments().delete(commentResponse.id).execute();
        assertSuccessfulResponse(response);
        assertThat(response.code()).isEqualTo(HttpURLConnection.HTTP_NO_CONTENT);
    }

    @Test
    public void test_replies() throws InterruptedException, IOException {
        // first post a new comment
        Comment comment = new Comment(buildTestEpisode(), "This is a toasty comment!", true, false);
        Comment response = executeCall(getTrakt().comments().post(comment));

        // give the server some time to handle the data
        Thread.sleep(3000);

        // post a reply to the new comment
        Comment expectedReply = new Comment("This is a reply to the toasty comment!", false, false);
        Comment actualReply = executeCall(getTrakt().comments().postReply(response.id, expectedReply));
        assertCommentResponse(expectedReply, actualReply);

        // give the server some time to handle the data
        Thread.sleep(3000);

        // look if the comment replies include our new reply
        List<Comment> replies = executeCall(getTrakt().comments().replies(response.id));
        for (Comment reply : replies) {
            if (reply.id.equals(actualReply.id)) {
                assertCommentResponse(actualReply, reply);
            }
        }

        // delete the comment and replies (does this work?)
        Response<Void> deleteResponse = getTrakt().comments().delete(response.id).execute();
        assertSuccessfulResponse(deleteResponse);
        assertThat(deleteResponse.code()).isEqualTo(HttpURLConnection.HTTP_NO_CONTENT);
    }

    private static void assertCommentResponse(Comment expected, Comment actual) {
        assertThat(actual.comment).isEqualTo(expected.comment);
        assertThat(actual.spoiler).isEqualTo(expected.spoiler);
        assertThat(actual.review).isEqualTo(expected.review);
    }

    private static Episode buildTestEpisode() {
        Episode episode = new Episode();
        episode.ids = EpisodeIds.tvdb(TestData.EPISODE_TVDB_ID);
        return episode;
    }
}

File: src/test/kotlin/com/redissi/trakt/services/CommentsTest.kt
package com.redissi.trakt.services

import com.redissi.trakt.BaseTestCase
import com.redissi.trakt.TestData
import com.redissi.trakt.TestResponse
import com.redissi.trakt.entities.Comment
import com.redissi.trakt.entities.Episode
import com.redissi.trakt.entities.EpisodeIds
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldNotBeNull
import org.junit.Test
import java.io.IOException
import java.net.HttpURLConnection

class CommentsTest : BaseTestCase(), TestResponse {
    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun `post and update a command`() = runBlocking<Unit> {
        // first post a new comment
        val comment = Comment.build(
            buildTestEpisode(),
            "This is a toasty comment!",
            spoiler = true,
            review = false
        )
        var commentResponse = trakt.comments().post(comment)

        commentResponse.shouldNotBeNull()
        commentResponse.id.shouldNotBeNull()

        // update the new comment
        val updatedComment = Comment.build("This is toasty! I was just updated.", spoiler = false, review = false)
        commentResponse = trakt.comments().update(commentResponse.id!!, updatedComment)

        commentResponse.shouldNotBeNull()

        assertCommentResponse(updatedComment, commentResponse)

        // delete the comment again
        val response = trakt.comments().delete(commentResponse.id!!)
        assertSuccessfulResponse(response, trakt)
        response.code().shouldBeEqualTo(HttpURLConnection.HTTP_NO_CONTENT)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun `delete a comment`() = runBlocking<Unit> {
        // first post a new comment
        val comment = Comment.build(
            buildTestEpisode(),
            "This is toasty! I should be deleted soon.",
            spoiler = true,
            review = false
        )
        val commentResponse = trakt.comments().post(comment)

        commentResponse.shouldNotBeNull()
        commentResponse.id.shouldNotBeNull()

        // delete the comment again
        val response = trakt.comments().delete(commentResponse.id!!)
        assertSuccessfulResponse(response, trakt)
        response.code().shouldBeEqualTo(HttpURLConnection.HTTP_NO_CONTENT)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun `reply to a comment`() = runBlocking<Unit> {
        // first post a new comment
        val comment = Comment.build(
            buildTestEpisode(),
            "This is a toasty comment!",
            spoiler = true,
            review = false
        )
        val response = trakt.comments().post(comment)
        response.shouldNotBeNull()
        response.id.shouldNotBeNull()
        // post a reply to the new comment
        val expectedReply = Comment.build("This is a reply to the toasty comment!", spoiler = false, review = false)
        val actualReply = trakt.comments().postReply(response.id!!, expectedReply)
        actualReply.shouldNotBeNull()
        actualReply.id.shouldNotBeNull()
        assertCommentResponse(expectedReply, actualReply)
        // look if the comment replies include our new reply
        val replies = trakt.comments().replies(response.id!!)
        replies.shouldNotBeNull()
        for (reply in replies) {
            if (reply.id == actualReply.id) {
                assertCommentResponse(actualReply, reply)
            }
        }
        // delete the comment and replies (does this work?)
        val deleteResponse = trakt.comments().delete(response.id!!)
        assertSuccessfulResponse(deleteResponse, trakt)
        deleteResponse.code().shouldBeEqualTo(HttpURLConnection.HTTP_NO_CONTENT)
    }

    companion object {
        private fun assertCommentResponse(
            expected: Comment,
            actual: Comment
        ) {
            actual.comment.shouldBeEqualTo(expected.comment)
            actual.spoiler.shouldBeEqualTo(expected.spoiler)
            actual.review.shouldBeEqualTo(expected.review)
        }

        private fun buildTestEpisode(): Episode {
            return Episode(ids = EpisodeIds.tvdb(TestData.EPISODE_TVDB_ID))
        }
    }
}